### PR TITLE
Fix filter parsing bug

### LINKED ISSUE #9507: [Bug] Semantic Layer where filter strings are parsed into lists
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Currently, if you pass a string into a filter YAML param, it will be assumed to be a list. This means your semantic manifest ends up with a list of strings for where filters. For example, this YAML:
```
filter: |
    {{ Dimension('customer__customer_type') }}  = 'new'
```
would look like this in the semantic manifest:
```
{"where_filters": [{"where_sql_template": "{"}, {"where_sql_template": "{"}, {"where_sql_template": " "}, {"where_sql_template": "D"}, {"where_sql_template": "i"}, {"where_sql_template": "m"}, {"where_sql_template": "e"}, {"where_sql_template": "n"}, {"where_sql_template": "s"}, {"where_sql_template": "i"}, {"where_sql_template": "o"}, {"where_sql_template": "n"}, {"where_sql_template": "("}, {"where_sql_template": "'"}, {"where_sql_template": "c"}, {"where_sql_template": "u"}, {"where_sql_template": "s"}, {"where_sql_template": "t"}, {"where_sql_template": "o"}, {"where_sql_template": "m"}, {"where_sql_template": "e"}, {"where_sql_template": "r"}, {"where_sql_template": "_"}, {"where_sql_template": "_"}, {"where_sql_template": "c"}, {"where_sql_template": "u"}, {"where_sql_template": "s"}, {"where_sql_template": "t"}, {"where_sql_template": "o"}, {"where_sql_template": "m"}, {"where_sql_template": "e"}, {"where_sql_template": "r"}, {"where_sql_template": "_"}, {"where_sql_template": "t"}, {"where_sql_template": "y"}, {"where_sql_template": "p"}, {"where_sql_template": "e"}, {"where_sql_template": "'"}, {"where_sql_template": ")"}, {"where_sql_template": " "}, {"where_sql_template": "}"}, {"where_sql_template": "}"}, {"where_sql_template": " "}, {"where_sql_template": " "}, {"where_sql_template": "="}, {"where_sql_template": " "}, {"where_sql_template": "'"}, {"where_sql_template": "n"}, {"where_sql_template": "e"}, {"where_sql_template": "w"}, {"where_sql_template": "'"}]}
```

### Expected Behavior

The semantic manifest should show this:
`{"where_filters": [{"where_sql_template": "{{ Dimension('customer__customer_type') }}  = 'new'"}]}`

### Steps To Reproduce

Add a string filter to a metric or saved query and run `dbt parse`, then look at the generated semantic manifest.

### Relevant log output

_No response_

### Environment

```markdown
- OS:
- Python:
- dbt:
```

### Which database adapter are you using with dbt?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by courtneyholcomb**
resolves #9507

### Problem
Currently, if you pass a string into a filter YAML param, it will be assumed to be a list. This means your semantic manifest ends up with a list of strings for where filters. For example, this YAML:
```
filter: |
    {{ Dimension('customer__customer_type') }}  = 'new'
```
would look like this in the semantic manifest:
```
{"where_filters": [{"where_sql_template": "{"}, {"where_sql_template": "{"}, {"where_sql_template": " "}, {"where_sql_template": "D"}, {"where_sql_template": "i"}, {"where_sql_template": "m"}, {"where_sql_template": "e"}, {"where_sql_template": "n"}, {"where_sql_template": "s"}, {"where_sql_template": "i"}, {"where_sql_template": "o"}, {"where_sql_template": "n"}, {"where_sql_template": "("}, {"where_sql_template": "'"}, {"where_sql_template": "c"}, {"where_sql_template": "u"}, {"where_sql_template": "s"}, {"where_sql_template": "t"}, {"where_sql_template": "o"}, {"where_sql_template": "m"}, {"where_sql_template": "e"}, {"where_sql_template": "r"}, {"where_sql_template": "_"}, {"where_sql_template": "_"}, {"where_sql_template": "c"}, {"where_sql_template": "u"}, {"where_sql_template": "s"}, {"where_sql_template": "t"}, {"where_sql_template": "o"}, {"where_sql_template": "m"}, {"where_sql_template": "e"}, {"where_sql_template": "r"}, {"where_sql_template": "_"}, {"where_sql_template": "t"}, {"where_sql_template": "y"}, {"where_sql_template": "p"}, {"where_sql_template": "e"}, {"where_sql_template": "'"}, {"where_sql_template": ")"}, {"where_sql_template": " "}, {"where_sql_template": "}"}, {"where_sql_template": "}"}, {"where_sql_template": " "}, {"where_sql_template": " "}, {"where_sql_template": "="}, {"where_sql_template": " "}, {"where_sql_template": "'"}, {"where_sql_template": "n"}, {"where_sql_template": "e"}, {"where_sql_template": "w"}, {"where_sql_template": "'"}]}
```
This is because the dataclass `from_dict()` method only works for unions if `Union` is the outermost type annotation. We had it nested in an `Optional` annotation, so this fixes that.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

**Review by tlento — COMMENTED** [review]
This seems reasonable to me based on my minuscule knowledge of Mashumaro (where the from_dict method apparently comes from), but I'm not a maintainer here so I'll let someone from the core team do the approving/merging.

File: core/dbt/contracts/graph/unparsed.py [hunk 1]
```diff
@@ -564,7 +564,7 @@ def __bool__(self):
 @dataclass
 class UnparsedMetricInputMeasure(dbtClassMixin):
     name: str
-    filter: Optional[Union[str, List[str]]] = None
+    filter: Union[Optional[str], Optional[List[str]]] = None
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply core/dbt/contracts/graph/unparsed.py:567]: Sure that works too! Just tested to confirm.

**Comment by Fatal1ty** [comment]
> Consequently, I have no idea what the resulting from_dict code is actually doing

You can see the generated code by enabling [`debug`](https://github.com/Fatal1ty/mashumaro?tab=readme-ov-file#debug-config-option) in the config parameter as described in the docs.

> why this was an issue

Currently, Union types are handled naively by trying to use a handler for each variant type in the loop. Values of type`str` are handled without modifications by default and that can be an issue in `Union`. The most reliable solution in this case would be to add a custom handler for `str` that would check the value type, for example, using `isinstance`. You can see an example [here](https://github.com/Fatal1ty/mashumaro/issues/42#issuecomment-1484059794). I understand that this imperfection might be confusing and dangerous in some cases, so I have plans to redo the processing of unions in next versions.

**Review by QMalcolm — CHANGES_REQUESTED** [review]
We should absolutely get this in. Thank you for doing this work ❤️ We should though add a test. Especially as @Fatal1ty mentioned, the typing based parser in Mashumaro will be seeing some changes in the future, so we want to guarantee that the filter continues to be parsed correctly as the things we depend on change how they operate. A good place to add a test would be in [test_metrics.py](https://github.com/dbt-labs/dbt-core/blob/main/tests/functional/metrics/test_metrics.py).

File: core/dbt/contracts/graph/unparsed.py [hunk 2]
```diff
@@ -564,7 +564,7 @@ def __bool__(self):
 @dataclass
 class UnparsedMetricInputMeasure(dbtClassMixin):
     name: str
-    filter: Optional[Union[str, List[str]]] = None
+    filter: Union[str, List[str], None] = None
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply core/dbt/contracts/graph/unparsed.py:568]: Good call, updated!

**Comment by courtneyholcomb** [comment]
Added tests! @QMalcolm thank you for pointing me to where those should be!

Note that in the process of writing tests, I discovered a separate bug. If you try to use a list filter that includes jinja on a metric or an input measure, you'll get an error when you run `dbt parse`. For example, parsing the following YAML:
```
metrics:
  - name: collective_tenure_measure_filter_list
    label: "Collective tenure2"
    description: Total number of years of team experience
    type: simple
    type_params:
      measure:
        name: "years_tenure"
        filter:
          - "{{ Dimension('id__loves_dbt') }} is true"
```
Gets this error:
`CompilationError("Could not render {{ Dimension('id__loves_dbt') }} is true: 'Dimension' is undefined")`
The exact same YAML works if you include the filter as a string.

I'll be putting up a separate issue for that bug. But that's the reason you don't see any test cases for input measures or metrics with list filters. The same bug does not exist for filters on saved queries, oddly, so I was able to include list filter tests for those.

**Comment by courtneyholcomb** [comment]
I'm not sure why the artifacts check is failing in CI - it looks like the action itself is erroring? But LMK if there's something I need to fix there!

**Comment by QMalcolm** [comment]
> I'm not sure why the artifacts check is failing in CI - it looks like the action itself is erroring? But LMK if there's something I need to fix there!

The failing CI check is one we've added recently. I've added the label `artifact_minor_upgrade` which essentially skips the check.

**Context**: we want to strongly control changes to `/artifacts` to help identify and mitigate breaking changes. To do this we added a CI step which checks for any file changes in `/artifacts`. Once we someone verifies the change isn't breaking, we add the `artifact_minor_upgrade` label to acknowledge this. We could implement a smarter check which recognizes what is and isn't a breaking change, and maybe we will in the future, but this gets the job done for now.

**Review by QMalcolm — APPROVED** [review]
Looks good to me! Thank you for adding the tests ❤️

**Pull request merged**